Näytä Välikatselmuksen päätökset kojelaudassa

jarnovayrynen opened the pull request.

jarnovayrynen committed:
Korjaa välikatselmuksen päätöksen metatiedot

-Luodessa älä aseta muokkausmetatietoja
-Poistaessa aseta muokkausmetatiedot

jarnovayrynen committed:
Näytä kojelaudassa Välikatselmuksen päätökset

jarnovayrynen committed:
Testaa että kojelaudassa näkyy Välikatselmuksen päätökset

jarnovayrynen committed:
Termimuutos päällystykseen hoitovuosi -> vuosi

jarnovayrynen committed:
Korjaa logiikka välikats. takarajan vertaamisessa

Esim 2023-2024 välikatselmuksen takaraja on 15.11.2024. Tässä oli off by one virhe.

jarnovayrynen committed:
Näytä välikatselmuksen tilojen yhteenveto

jarnovayrynen committed:
Lisää taoenssa clj-kondot

jarnovayrynen committed:
Merge branch 'develop' of github.com:finnishtransportagency/harja into HARJA-985

jarnovayrynen committed:
Merge branch 'develop' of github.com:finnishtransportagency/harja into HARJA-985

solita-staskila commented on src/cljs/harja/views/hallinta/kojelauta.cljs:
```diff
@@ -108,11 +143,13 @@
     :rivi-jalkeen-fn (fn [urakat]
-                       (let [ks-tilojen-yhteenveto (tiedot/ks-tilojen-yhteenveto urakat)]
+                       (let [ks-tilojen-yhteenveto (tiedot/ks-tilojen-yhteenveto urakat)
+                             valikatselmus-tilojen-yhteenveto (tiedot/valikatselmus-tilojen-yhteenveto urakat)]
```
Oisko gridille `:sivuta 20` esim, ja jos hoitovuodelle annettais leveydeksi 5, tulee tällaisella vähän pienemmälläkin resolla nätimmän näköinen: 

![image](https://github.com/user-attachments/assets/2fdbedcc-2fdd-4352-b0a1-37996bd495ab)

jarnovayrynen replied:
Laitoin ton 20 rivin sivutuksen, vois olla ihan hyvä idea. On se 80 urakan lista aika pitkä rullattavaksi. Pitää sitten tarkkaan miettiä, toimiiko se myös korjauspuolella, jossa mahdollisesti alle 30 urakkaa, jolloin sivutuksesta voi olla enemmän haittaa, ja voimme tehdä kynnysarvosta dynaamisen urakkatyypin perusteella.

solita-staskila commented on src/cljs/harja/tiedot/hallinta/kojelauta.cljs:
```diff
@@ -28,6 +28,22 @@
+        urakat-joissa-ei-paatoksia (- kaikkien-urakoiden-lkm urakat-joissa-jokin-rahapaatos-tehtyna urakat-joissa-jokin-lupauspaatos-tehtyna)
+        valikatselmusten-yhteenveto (when-not (empty? urakat)
+                                      [:span.valikatselmustiedot
+                                       [:div (str "Ei yhtään päätöstä: " urakat-joissa-ei-paatoksia " (" (fmt/prosentti-opt (* 100 (/ urakat-joissa-ei-paatoksia kaikkien-urakoiden-lkm))) ")")]
```
Tähän tulee nyt -1: 

![image](https://github.com/user-attachments/assets/8c640fc0-3e50-40a8-ad69-322bd9defc2b)


Tässä muuttujien arvot, eli nämä vaan kaikki miinustetaan, tulee -1:
```
kaikkien-urakoiden-lkm  1
urakat-joissa-jokin-rahapaatos-tehtyna  1
urakat-joissa-jokin-lupauspaatos-tehtyna  1 
```

jarnovayrynen replied:
Joo tässähän on täydellinen ajatusvirhe, hyvä kun huomasit. :) Korjaan.

solita-staskila commented on src/cljs/harja/views/hallinta/kojelauta.cljs:
```diff
@@ -99,7 +134,7 @@
 
 (defn taulukko-paallystysurakat [e!  {:keys [urakat haku-kaynnissa?]}]
-  [:div "Tänne päällystysurakoiden taulukko samaan tyyliin kuin alla hoitourakoille..."])
+  [:div "Tänne tulee lähiaikoina päällystysurakoiden tietoa..."])
```
Kun urakkaa hakee, vaikka yllä oleva imatra, tuota ei saa kumottua enää mitenkään. Tarvitaan jokin nappi, tai muu toiminto resetoimaan suodattimet ilman että sivua tarvitsee uudelleenladata. (Puhun tuosta olemassa olevasta hoitourakoiden taulukoista)

solita-staskila commented on src/cljs/harja/views/hallinta/kojelauta.cljs:
```diff
@@ -74,13 +75,47 @@
   (let [indeksi-saatavilla? (boolean (:indeksikerroin rivi))
         {:keys [aloittamattomia vahvistamattomia vahvistettuja suunnitelman_tila]} (:ks_tila rivi)]
     [yleiset/wrap-if true
      [yleiset/tooltip {} :% "Siirry kustannussuunnitelmaan"]
```
Välikatselmus- sarakkeessa, kustiksen sijaan, me voitaisiin siirtyä suoraan välikatselmukseen. 

Tämä kuitenkin vaati hieman virvelöintiä, koska meidän täytyy ensin odottaa että valittu urakka on ladannut ennenkuin voidaan siirtyä välikatselmukseen, jotenka, tässä on toimiva koodi jonka voit halutessas lisätä:

<details>
<summary> ᴄʟɪᴄᴋ ᴛᴏ ᴇxᴘᴀɴᴅ </summary>

```clojure 

[yleiset/wrap-if true
     [yleiset/tooltip {} :% "Siirry välikatselmukseen"]
     [:a.klikattava {:on-click #(do
	  ;; Valitse urakka
	  (siirtymat/kustannusten-seurantaan-valitussa-urakassa (:ely_id rivi) (:id rivi))
	  ;; Pollaa kunnes valittu urakka id on klikattu urakka ^ 
	  (let [max-yritykset 20
			yritykset (atom 0)
			interval-id (atom nil)]

		(reset! interval-id
		  (js/setInterval
			(fn []
			  (println "\n __poll...")

			  ;; Nosta yritysmäärää
			  (swap! yritykset inc)

			  ;; Odota, että valittu urakka on ladannut 
			  (if (= (:id @nav/valittu-urakka) (:id rivi))
				(do
				  ;; Urakka on ladannut -> clearInterval -> avaa-valikatselmus
				  (println "__call: löytyi id: " (:id rivi) "  valittuun urakkaan: " (:id @nav/valittu-urakka))
				  (js/clearInterval @interval-id)
				  (.setTimeout js/window
							   (fn []
								 (siirtymat/avaa-valikatselmus
								   ;; Palauttaa halutun välikatselmuksen aikavälin, 
								   ;; esim: [#object[Object 20221001T000000] #object[Object 20230930T235959]]
								   (nth (urakka/hoito-tai-sopimuskaudet @nav/valittu-urakka)
									 (dec (urakka/hoitokauden-jarjestysnumero hoitokauden_alkuvuosi
											(-> @tila/yleiset :urakka :loppupvm))))))
							   1800))
				;; Lopeta pollaus jos tuloksia ei ikinä tule 
				(when (>= @yritykset max-yritykset)
				  (println "__call: max yritysmäärä saavutettiin -> clearInterval")
				  (js/clearInterval @interval-id))))
			550))) ;; Pollaa joka 0.5s
	  )}

```

</details>

&nbsp;

⚠️ HOX : 
Tässä on Tonilla tällä hetkellä työn alla bugikorjaus, jossa **hoitokauden-jarjestysnumero** ei toimi oikein urakoille joilla 6 hoitokautta (tällä hetkellä ongelma on pelkästään toi imatran urakka id: 458), tähän varmaan tulossa sieltäpäin korjaus. Kaikille muille urakoille ton pitäis toimia.

solita-staskila commented on src/clj/harja/kyselyt/valikatselmus.clj:
```diff
@@ -112,6 +112,7 @@
           ::muokkaustiedot/poistettu? false}))
 
 (defn tee-paatos [db paatos]
+  (prn "Jarno tee päätös: " paatos)
```
Tää varmaan jäänyt?

solita-staskila reviewed (approved):
Tuo "Hae urakkaa" valinta pitäis saada jotenkin resetoitua helpommin. Yhteenvedossa yks numero näyttää -1. Noi kun korjaat niin hyvä. 

Muut jutut tais olla vaan ehdotuksia. Hyvin tää pelaa.

jarnovayrynen committed:
Korjaa välikatselmustietoja kommenttien pohjalta

Korjaa logiikkavirhe niiden urakoiden laskennassa, joissa ei ole tehty mitään päätöstä.
Lisää sivutus.
